internal/olm/operator/registry/operator_installer_test.go

```diff
@@ -16,22 +16,233 @@ package registry
+
+			sub := newSubscription(oi.StartingCSV, oi.cfg.Namespace,
+				withPackageChannel(oi.PackageName, oi.Channel, oi.StartingCSV),
+				withCatalogSource("duplicate", oi.cfg.Namespace),
+				withInstallPlanApproval(v1alpha1.ApprovalManual))
+			oi.cfg.Client = fake.NewFakeClientWithScheme(sch, sub)
+
+			_, err := oi.createSubscription(context.TODO(), "duplicate")
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).Should(ContainSubstring("error creating subscription"))
+		})
+	})
+
+	// TODO(asmacdo) Since this uses the olm client can/should this be unit tested?
+	// Describe("getInstalledCSV", func() {
+	// 	})
+	// })
+	Describe("approveInstallPlan", func() {
+		var (
+			oi  *OperatorInstaller
+			sch *runtime.Scheme
+		)
+		BeforeEach(func() {
+			kubeconfigPath := os.Getenv(k8sutil.KubeConfigEnvVar)
+			cfg := &operator.Configuration{KubeconfigPath: kubeconfigPath}
+			sch = runtime.NewScheme()
+			Expect(v1alpha1.AddToScheme(sch)).To(Succeed())
+			oi = NewOperatorInstaller(cfg)
+		})
+
+		It("should update the install plan", func() {
+			existingIp := &v1alpha1.InstallPlan{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "fakeName",
+					Namespace: "fakeNS",
+				},
+			}
+			oi.cfg.Client = fake.NewFakeClientWithScheme(sch, existingIp)
+
+			ip := &v1alpha1.InstallPlan{}
+			ipKey := types.NamespacedName{
+				Namespace: "fakeNS",
+				Name:      "fakeName",
+			}
+
+			err := oi.cfg.Client.Get(context.TODO(), ipKey, ip)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(ip.Name).To(Equal("fakeName"))
+			Expect(ip.Namespace).To(Equal("fakeNS"))
+			// Sanity Check
+			Expect(ip.Spec.Approved).To(Equal(false))
+
+			// Test
+			sub := &v1alpha1.Subscription{
+				Status: v1alpha1.SubscriptionStatus{
+					InstallPlanRef: &corev1.ObjectReference{
+						Name:      "fakeName",
+						Namespace: "fakeNS",
+					},
+				},
+			}
+			err = oi.approveInstallPlan(context.TODO(), sub)
+			Expect(err).ToNot(HaveOccurred())
+			err = oi.cfg.Client.Get(context.TODO(), ipKey, ip)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(ip.Name).To(Equal("fakeName"))
+			Expect(ip.Namespace).To(Equal("fakeNS"))
+			Expect(ip.Spec.Approved).To(Equal(true))
+		})
+		It("should return an error if the install plan does not exist.", func() {
+			oi.cfg.Client = fake.NewFakeClientWithScheme(sch)
+			sub := &v1alpha1.Subscription{
+				Status: v1alpha1.SubscriptionStatus{
+					InstallPlanRef: &corev1.ObjectReference{
+						Name:      "fakeName",
+						Namespace: "fakeNS",
+					},
+				},
+			}
+			err := oi.approveInstallPlan(context.TODO(), sub)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).Should(ContainSubstring("error getting install plan"))
+		})
+		// TODO(asmacdo) How to throw an err on update?
+		// It("should return an error if the install plan fails to update.", func() {
+		// 	Expect(err.Error()).Should(ContainSubstring("error approving install plan"))
+		// })
```

Comment:
You could test one scenario by having an EMPTY fake client:

```golang
    if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
        if err := o.cfg.Client.Get(ctx, ipKey, &ip); err != nil {
            return fmt.Errorf("error getting install plan: %v", err)
        }
        ...
```
If you use an empty fakeclient the get will fail with `NotFound` error. Then you can validate that the error contains "error getting install plan".

The `Update` part of the code is harder, not sure yet the best way to do that.

Reply:
I tested the empty client 1 test up. (Do we prefer HappyPath tests before err tests?)